node-manager/src/socket.rs

```diff
@@ -0,0 +1,413 @@
+impl PartialEq for WsTcpSocketDescriptor {
+    fn eq(&self, o: &Self) -> bool {
+        self.id == o.id
+    }
+}
+impl Hash for WsTcpSocketDescriptor {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.id.hash(state);
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct MultiWsSocketDescriptor {
+    conn: Arc<Proxy>,
+    read_from_sub_socket: Receiver<Message>,
+    send_to_multi_socket: Sender<Message>,
+    socket_map: Arc<Mutex<HashMap<Vec<u8>, Sender<Message>>>>,
+    peer_manager: Arc<PeerManager>,
+    connected: Arc<AtomicBool>,
+}
+impl MultiWsSocketDescriptor {
+    pub fn new(conn: Arc<Proxy>, peer_manager: Arc<PeerManager>) -> Self {
+        debug!("setting up multi websocket descriptor");
+
+        let (send_to_multi_socket, read_from_sub_socket): (Sender<Message>, Receiver<Message>) =
+            unbounded();
+
+        let socket_map: Arc<Mutex<HashMap<Vec<u8>, Sender<Message>>>> =
+            Arc::new(Mutex::new(HashMap::new()));
+
+        Self {
+            conn,
+            send_to_multi_socket,
+            read_from_sub_socket,
+            socket_map,
+            peer_manager,
+            connected: Arc::new(AtomicBool::new(true)),
+        }
+    }
+
+    pub fn connected(&self) -> bool {
+        self.connected.load(Ordering::Relaxed)
+    }
+
+    pub fn reconnect(&mut self, conn: Arc<Proxy>) {
+        let (send_to_multi_socket, read_from_sub_socket): (Sender<Message>, Receiver<Message>) =
+            unbounded();
+
+        let socket_map: Arc<Mutex<HashMap<Vec<u8>, Sender<Message>>>> =
+            Arc::new(Mutex::new(HashMap::new()));
+
+        self.conn = conn;
+        self.send_to_multi_socket = send_to_multi_socket;
+        self.read_from_sub_socket = read_from_sub_socket;
+        self.socket_map = socket_map;
+        self.connected.store(true, Ordering::Relaxed);
+
+        self.listen();
+    }
+
+    pub async fn create_new_subsocket(&self, id: Vec<u8>) -> SubWsSocketDescriptor {
+        create_new_subsocket(
+            self.socket_map.clone(),
+            self.send_to_multi_socket.clone(),
+            id,
+        )
+        .await
+    }
+
+    pub fn listen(&self) {
+        let conn_copy = self.conn.clone();
+        let socket_map_copy = self.socket_map.clone();
+        let send_to_multi_socket_copy = self.send_to_multi_socket.clone();
+        let peer_manager_copy = self.peer_manager.clone();
+        let connected_copy = self.connected.clone();
+        debug!("spawning multi socket connection reader");
+        spawn_local(async move {
+            while let Some(msg) = conn_copy.read.lock().await.next().await {
+                if let Ok(Message::Bytes(msg)) = msg {
+                    // parse the msg and see which pubkey it belongs to
+                    if msg.len() < PUBKEY_BYTES_LEN {
+                        debug!("msg not long enough to have pubkey, ignoring...");
+                        continue;
+                    }
```

Comment:
Could this just be we need to wait for more bytes? Might need to implement some buffering, probably fine for now if you got payments working

Reply:
I don't think this should ever happen. Buffering shouldn't be needed with these websocket libraries, it should handle that.